catroid/src/main/java/org/catrobat/catroid/content/actions/ChangeVariableAction.kt

```diff
@@ -0,0 +1,58 @@
+/*
+ * Catroid: An on-device visual programming system for Android devices
+ * Copyright (C) 2010-2018 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.catrobat.catroid.content.actions
+
+import android.util.Log
+import com.badlogic.gdx.scenes.scene2d.Action
+import org.catrobat.catroid.content.Sprite
+import org.catrobat.catroid.formulaeditor.Formula
+import org.catrobat.catroid.formulaeditor.UserVariable
+
+class ChangeVariableAction : Action() {
+    var sprite: Sprite? = null
+    var changeVariable: Formula? = null
+    var userVariable: UserVariable? = null
+
+    override fun act(delta: Float): Boolean {
+        val originalValue = userVariable?.value ?: return true
+        if (originalValue is String) {
+            return true
+        }
+        var value = changeVariable?.interpretObject(sprite) ?: 0.0
+        if (value is String) {
+            value = try { java.lang.Double.valueOf(value) } catch (e: NumberFormatException) {
+                Log.d(javaClass.simpleName, "Couldn't parse String", e)
+            }
+        }
+        updateUserVariable(originalValue, value)
```

Comment:
You don't need to do it exactly like this, just a suggestion. This also checks for double and allows you to remove the type check in updateUserVariable(...) -> see below

```suggestion
        (value as? Double ?: (value as? String)?.toDoubleOrNull())?.run {
            updateUserVariable(originalValue, this)
        }
```